a2s/a2s_async.py

```diff
@@ -1,138 +1,226 @@
+from __future__ import annotations
+
 import asyncio
+import io
 import logging
 import time
-import io
+from typing import TYPE_CHECKING, Dict, List, NoReturn, Optional, Tuple, Type, TypeVar, Union, overload
 
-from a2s.exceptions import BrokenMessageError
-from a2s.a2s_fragment import decode_fragment
-from a2s.defaults import DEFAULT_RETRIES
+from a2s.a2s_fragment import A2SFragment, decode_fragment
 from a2s.byteio import ByteReader
+from a2s.defaults import DEFAULT_RETRIES
+from a2s.exceptions import BrokenMessageError
 
+from .info import GoldSrcInfo, InfoProtocol, SourceInfo
+from .players import Player, PlayersProtocol
+from .rules import RulesProtocol
 
+if TYPE_CHECKING:
+    from typing_extensions import Self
 
 HEADER_SIMPLE = b"\xFF\xFF\xFF\xFF"
 HEADER_MULTI = b"\xFE\xFF\xFF\xFF"
 A2S_CHALLENGE_RESPONSE = 0x41
+PROTOCOLS = Union[InfoProtocol, PlayersProtocol, RulesProtocol]
+
+logger: logging.Logger = logging.getLogger("a2s")
+
+T = TypeVar("T", InfoProtocol, PlayersProtocol, RulesProtocol)
+
 
-logger = logging.getLogger("a2s")
+@overload
+async def request_async(
+    address: Tuple[str, int], timeout: float, encoding: str, a2s_proto: Type[InfoProtocol]
+) -> Union[SourceInfo, GoldSrcInfo]:
+    ...
 
 
-async def request_async(address, timeout, encoding, a2s_proto):
+@overload
+async def request_async(
+    address: Tuple[str, int], timeout: float, encoding: str, a2s_proto: Type[PlayersProtocol]
+) -> List[Player]:
+    ...
```

Comment:
`a2s_proto` typing needs to be done with inheritance from a base type, this module should not need to know about the functions calling it.

Reply:
Ideally I would agree, but since this type is what allows the overloads to denote the return type of the function I'm unsure how else to handle this from a library side.
Currently I believe the options are:-

1. Return merely a `Union` type of all 3(?) possible return types and let users handle it in usercode.
2. Add another argument for an enum which would control the type checking's return type information via overloads (although this sounds a bit clunky)...

If you have another suggestion I'd be very happy to hear it!

Reply:
The only way this could be implemented in a useful way would be to inherit the return type from `a2s_proto.deserialize_response`. If that's not possible, it should just have `Any` return type and that's okay.